src/app/services/ChatService.ts

```diff
@@ -72,26 +49,58 @@ export class ChatService implements OnDestroy {
     }
     public isObserving(): boolean {
         return this.followedChatId != null;
     }
-    public async deleteChat(chatId: string): Promise<void> {
+    public async deleteChat(chatId: NonNullable<string>): Promise<void> {
         display(ChatService.VERBOSE, 'ChatService.deleteChat ' + chatId);
 
-        if (chatId == null) {
-            throw new Error('Cannot delete chat of null id');
-        }
+        assert(chatId != null, 'cannot delete chat of null id');
         return this.chatDao.delete(chatId);
     }
-    // delegate
-
-    public async set(id: string, chat: IChat): Promise<void> {
-        return this.chatDao.set(id, chat);
+    public async createNewChat(id: string): Promise<void> {
+        return this.chatDao.set(id, {
+            messages: [],
+        });
+    }
+    public async sendMessage(userName: string, currentTurn: number, content: string): Promise<MGPValidation> {
+        if (this.userCanSendMessage(userName, this.followedChatId) === false) {
+            return MGPValidation.failure(ChatMessages.CANNOT_SEND_MESSAGE);
+        }
+        if (this.isForbiddenMessage(content)) {
+            return MGPValidation.failure(ChatMessages.FORBIDDEN_MESSAGE);
+        }
+        const chat: IChat = await this.chatDao.read(this.followedChatId);
+        const messages: IMessage[] = ArrayUtils.copyImmutableArray(chat.messages);
+        const newMessage: IMessage = {
+            content,
+            sender: userName,
+            postedTime: Date.now(),
+            currentTurn,
+        };
+        messages.push(newMessage);
+        await this.chatDao.update(this.followedChatId, { messages });
+        return MGPValidation.SUCCESS;
+    }
+    private userCanSendMessage(userName: string, chatId: string): boolean {
+        // TODO FOR REVIEW: should check that user is part of the game or is an observer
+        // Quentin: it seems that this should be encoded in the security rules, not in the application logic.
+        // If it is not encoded in the security rules, it is possible to craft firebase queries to send messages where we're not supposed to.
+        // And if it is encoded, there is duplication between the app logic and the security rules.
+        // Also, checking it in the app is useless in the sense that userCanSendMessage *should* never be false: if we reach sendMessage, this is by doing actions that the user can.
+        // The best IMHO is: encode it in the security rules, catch "write" failures to firebase in DAO, and propagate that back
+        // If we agree, we can remove this check and the corresponding TODO.
+        return userName != null;
+    }
+    private isForbiddenMessage(message: string): boolean {
+        // TODO FOR REVIEW: improve? what should be forbidden in a message?
+        // Quentin: I think here the only check should be indeed for '', but nothing else.
+        // I don't see a reason to forbid anything else.
+        // If we agree, we can remove the TODO.
```

Comment:
might refuse flood, message containing insult, or like, people saying "j'ai perdu!" :D

Reply:
ok pour le flood! Insulte et autre tu les distingue comment ? Une liste de mots ? Donc je peux dire f*ck you ? (La plupart des chats que j'ai vu qui empêchent les insultent les empêchent qu'en surface :D)